### PR TITLE
OCaml: default to version 4.09

### PR DESCRIPTION
**Pull request by vbgl**
###### Motivation for this change

nixpkgs is ready for 4.09

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/ocaml-packages.nix [hunk 1]
```diff
@@ -1232,5 +1232,5 @@ in let inherit (pkgs) callPackage; in rec
 
   ocamlPackages_latest = ocamlPackages_4_10;
 
-  ocamlPackages = ocamlPackages_4_08;
+  ocamlPackages = ocamlPackages_4_09;
```
Comment (Mic92) [inline pkgs/top-level/ocaml-packages.nix]: What are our policies for keeping version of the ocaml compiler though?
Which of these version are still maintained upstream?

**Comment by vbgl** [comment]
> What are our policies for keeping version of the ocaml compiler though?

IMHO, old compilers should be removed when maintaining them becomes a
burden.

> Which of these version are still maintained upstream?

I don’t know.

**Comment by Mic92** [comment]
> What are our policies for keeping version of the ocaml compiler though?
> IMHO, old compilers should be removed when maintaining them becomes a burden.
> Which of these version are still maintained upstream?
> I don’t know.

I suppose since we don't build all ocaml package sets this is fine.

**Comment by vbgl** [comment]
@GrahamcOfBorg build acgtk beluga bibtex2html coq2html coq_8_10 eff flow framac hol_light llpp obelisk ocaml-top opam ott patdiff pdfsandwich prooftree proverif reason satallax satysfi stog unison usync why3

**Comment by orivej** [comment]
ocaml-gettext needs to be updated to unbreak vagrant via libguestfs. (0.3.8 depends on camlp4 which [does not support 4.09](https://github.com/camlp4/camlp4/pull/153/commits/ea05c68d2de239d904c5c5288e67efc0ef2dd952), 0.4.0 [says](https://github.com/gildor478/ocaml-gettext/releases) that it has switched from camlp4 to ppx.)

**Comment by vbgl** [comment]
Thanks for the report. Sorry for breaking vagrant without noticing it.

I’ll give a try to the update ASAP.

**Comment by orivej** [comment]
Thank you! (In case updating ocaml-gettext breaks other things, the [patch](https://github.com/camlp4/camlp4/pull/153/commits/ea05c68d2de239d904c5c5288e67efc0ef2dd952) to make the current version support 4.09 is trivial.)

**Comment by vbgl** [comment]
Here is a PR updating ocaml-gettext:  #84457

**Pull request merged**